staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,570 @@
+	qs.numRequestsEnqueued++
+
+	metrics.UpdateFlowControlRequestsInQueue(qs.config.Name, qs.numRequestsEnqueued)
+}
+
+// rejectOrEnqueueLocked rejects or enqueues the newly arrived request if
+// resource criteria isn't met
+func (qs *queueSet) rejectOrEnqueueLocked(request *fq.Request) bool {
+	queue := request.Queue
+	curQueueLength := len(queue.Requests)
+	// rejects the newly arrived request if resource criteria not met
+	if qs.getRequestsExecutingLocked() >= qs.config.ConcurrencyLimit &&
+		curQueueLength >= qs.config.QueueLengthLimit {
+		return false
+	}
+
+	qs.enqueue(request)
+	return true
+}
+
+func (qs *queueSet) roundRobinQueue() int {
+	qs.robinIndex = (qs.robinIndex + 1) % len(qs.queues)
+	return qs.robinIndex
+}
+
+// selectQueue selects the minimum virtualFinish time from the set of queues
+// the starting queue is selected via roundrobin
+func (qs *queueSet) selectQueue() *fq.Queue {
+	minVirtualFinish := math.Inf(1)
+	var minQueue *fq.Queue
+	var minIndex int
+	for range qs.queues {
+		index := qs.roundRobinQueue()
+		queue := qs.queues[index]
+		if len(queue.Requests) != 0 {
+			currentVirtualFinish := queue.GetVirtualFinish(0, qs.estimatedServiceTime)
+			if currentVirtualFinish < minVirtualFinish {
+				minVirtualFinish = currentVirtualFinish
+				minQueue = queue
+				minIndex = index
+			}
+		}
+	}
+	qs.robinIndex = minIndex
+	return minQueue
+}
+
+// dequeue dequeues a request from the queueSet
+func (qs *queueSet) dequeue() (*fq.Request, bool) {
+	queue := qs.selectQueue()
+	if queue == nil {
+		return nil, false
+	}
+	request, ok := queue.Dequeue()
+	if !ok {
+		return nil, false
+	}
+	// When a request is dequeued for service -> qs.VirtualStart += G
+	queue.VirtualStart += qs.estimatedServiceTime
+	request.StartTime = qs.clock.Now()
+	// request dequeued, service has started
+	queue.RequestsExecuting++
+	metrics.UpdateFlowControlRequestsExecuting(qs.config.Name, queue.RequestsExecuting)
+	qs.numRequestsEnqueued--
+	return request, ok
+}
+
+// dequeueWithChannelAsMuchAsPossible runs a loop, as long as there
+// are non-empty queues and the number currently executing is less than the
+// assured concurrency value.  The body of the loop uses the fair queuing
+// technique to pick a queue, dequeue the request at the head of that
+// queue, increment the count of the number executing, and send true
+// to the request's channel.
+func (qs *queueSet) dequeueWithChannelAsMuchAsPossible() {
+	for qs.numRequestsEnqueued != 0 && qs.getRequestsExecutingLocked() < qs.config.ConcurrencyLimit {
+		_, ok := qs.dequeueWithChannel()
+		if !ok {
+			break
+		}
+	}
+}
+
+// dequeueWithChannel is convenience method for dequeueing requests that
+// require a message to be sent through the requests channel
+// this is a required pattern for the QueueSet the queueSet supports
+func (qs *queueSet) dequeueWithChannel() (*fq.Request, bool) {
+	req, ok := qs.dequeue()
+	if !ok {
+		return nil, false
+	}
+	qs.counter.Add(1)
```

Comment:
Why shouldn't the thread reading from the channel do this?

If the answer is some variation of "locking" then why shouldn't this be done by where queue.RequestsExecuting and qs.numRequestsEnqueued are updated in `dequeue`?

Reply:
By "this" you mean increment the counter?  That is for the same reason that the counter is incremented before forking a goroutine: doing it after might be too late.